File: backend/routers/pokemon.py
from fastapi import APIRouter, Query

from models.schemas import PaginatedResponse, PokemonDetail, PokemonSummary, EvolutionNode
from services import pokeapi_client

router = APIRouter(prefix="/api", tags=["pokemon"])


@router.get("/pokemon", response_model=PaginatedResponse)
async def list_pokemon(
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    type: str | None = Query(None),
):
    if type:
        return await pokeapi_client.get_pokemon_by_type(type, page, limit)
    return await pokeapi_client.list_pokemon(page, limit)


@router.get("/search", response_model=list[PokemonSummary])
async def search_pokemon(q: str = Query(..., min_length=1), limit: int = Query(10, ge=1, le=50)):
    return await pokeapi_client.search_pokemon(q, limit)


@router.get("/pokemon/{name_or_id}", response_model=PokemonDetail)
async def get_pokemon(name_or_id: str):
    return await pokeapi_client.get_pokemon(name_or_id.lower())


@router.get("/pokemon/{name}/evolution", response_model=list[EvolutionNode])
async def get_evolution(name: str):
    return await pokeapi_client.get_evolution(name.lower())

File: frontend/src/api/pokemon.ts
import api from "./client";
import type {
  PaginatedResponse,
  PokemonDetail,
  PokemonSummary,
  EvolutionNode,
  TypeInfo,
} from "../types/pokemon";

export async function fetchPokemons(
  page: number,
  limit: number,
  type?: string
): Promise<PaginatedResponse> {
  const params: Record<string, string | number> = { page, limit };
  if (type) params.type = type;
  const { data } = await api.get<PaginatedResponse>("/pokemon", { params });
  return data;
}

export async function fetchPokemon(
  nameOrId: string
): Promise<PokemonDetail> {
  const { data } = await api.get<PokemonDetail>(`/pokemon/${nameOrId}`);
  return data;
}

export async function fetchEvolution(
  name: string
): Promise<EvolutionNode[]> {
  const { data } = await api.get<EvolutionNode[]>(
    `/pokemon/${name}/evolution`
  );
  return data;
}

export async function fetchTypes(): Promise<TypeInfo[]> {
  const { data } = await api.get<TypeInfo[]>("/types");
  return data;
}

export async function searchPokemon(
  query: string
): Promise<PokemonSummary[]> {
  const { data } = await api.get<PokemonSummary[]>("/search", {
    params: { q: query },
  });
  return data;
}
